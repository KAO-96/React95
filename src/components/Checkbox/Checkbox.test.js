import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import PropTypes from "prop-types";
import { assertPropTypes } from "check-prop-types";

import Checkbox, { StyledCheckmark } from "./Checkbox";
import { padding, fontSizes } from "../common/system";
import * as common from "../common"

describe("Component: Checkbox", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  })

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  })

  const testProps = {
    onChange: () => ({}),
    name: "",
    value: "",
    label: "",
    checked: false,
    disabled: false,
    shadow: false,
    style: { color: "red" }
  };

  it("warns if props are missing or the wrong type", () => {
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, onChange: "myFunc" }, "prop", Checkbox.onChange); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, name: undefined }, "prop", Checkbox.name); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, value: undefined }, "prop", Checkbox.value); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, label: true }, "prop", Checkbox.label); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, checked: "true" }, "prop", Checkbox.checked); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, disabled: "true" }, "prop", Checkbox.disabled); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, shadow: "true" }, "prop", Checkbox.shadow); }).toThrow()
    expect(() => { assertPropTypes(Checkbox.propTypes, { ...testProps, shadow: "red" }, "prop", Checkbox.shadow); }).toThrow()
  })

  it("renders", () => {
    const wrapper = shallow(<Checkbox />);
    expect(wrapper.length).toBe(1);
  })

  it("renders className", () => {
    const wrapper = shallow(<Checkbox className="my-class" />);
    expect(wrapper.props().className).toBe("my-class");
  });

  it("renders style object", () => {
    const wrapper = shallow(<Checkbox style={{ color: "blue" }} />);
    expect(wrapper.props().style.color).toBe("blue");
  });

  it("renders other props passed", () => {
    const wrapper = shallow(<Checkbox a={1} b={2} c={3} />);
    const input = wrapper.childAt(0);
    expect(input.props().a).toBe(1);
    expect(input.props().b).toBe(2);
    expect(input.props().c).toBe(3);
  });

  it("renders default component styles", () => {
    common.createDisabledTextStyles = jest.fn();
    const tree = renderer.create(<Checkbox />).toJSON();
    expect(common.createDisabledTextStyles.mock.calls.length).toBe(0);
  });

  it("renders disabled component styles", () => {
    common.createDisabledTextStyles = jest.fn();
    const tree = renderer.create(<Checkbox disabled />).toJSON();
    expect(common.createDisabledTextStyles.mock.calls.length).not.toBe(0);
  });

  it("renders uncontrolled checkbox updates", () => {
    act(() => {
      ReactDOM.render(<Checkbox />, container);
    })

    const input = container.querySelector("input");
    expect(input.checked).toBe(false);

    act(() => {
      input.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    })

    expect(input.checked).toBe(true);
  })

  it("renders styledCheckmark with no shadow", () => {
    const tree = renderer.create(<StyledCheckmark />).toJSON();
    expect(tree).toHaveStyleRule("box-shadow", "none");
  })

  it("renders calls to onChange from uncontrolled checkboxes", () => {
    const onChange = jest.fn();
    const mockEvent = { target: { checked: true } }
    const wrapper = shallow(<Checkbox onChange={onChange} />);
    wrapper.find("[type='checkbox']").simulate("change", mockEvent);
    expect(onChange.mock.calls[0][0]).toEqual(mockEvent);
  })

  it("renders controlled checkbox updates", () => {
    let input;

    act(() => {
      ReactDOM.render(<Checkbox checked={false} />, container);
    })

    input = container.querySelector("input");
    expect(input.checked).toBe(false);

    act(() => {
      ReactDOM.render(<Checkbox checked={true} />, container);
    })

    input = container.querySelector("input");
    expect(input.checked).toBe(true);
  })

  it("renders inactive when disabled from a controlled component", () => {
    const wrapper = shallow(<Checkbox checked={true} disabled />);
    expect(wrapper.find("[type='checkbox']").props().onChange).toBeUndefined();
  })

  it("renders a matching snapshot", () => {
    const tree = renderer.create(<Checkbox value="foo" label="bar" />).toJSON();
    expect(tree).toMatchSnapshot();
  })
})
