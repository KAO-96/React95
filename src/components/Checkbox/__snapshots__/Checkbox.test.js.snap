// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Checkbox renders a matching snapshot 1`] = `
.c3 {
  position: relative;
  box-sizing: border-box;
  padding: 2px;
  border-style: solid;
  border-width: 2px;
}

.c3:before {
  position: absolute;
  left: 0;
  top: 0;
  z-index: 1;
  content: "";
  width: calc(100% - 4px);
  height: calc(100% - 4px);
  border-style: solid;
  border-width: 2px;
  pointer-events: none;
  box-shadow: inset 3px 3px 10px rgba(0,0,0,0.2);
}

.c0 {
  display: inline-block;
  position: relative;
  padding-left: calc(20px + 0.5rem);
  margin: 0.5rem 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 16px;
}

.c1 {
  position: absolute;
  opacity: 0;
  z-index: -99;
}

.c2 {
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 20px;
  height: 20px;
  box-shadow: inset 3px 3px 10px rgba(0,0,0,0.1);
}

.c2:before {
  box-shadow: none;
}

<label
  className="c0"
>
  bar
  <input
    checked={false}
    className="c1"
    name=""
    onChange={[Function]}
    readOnly={false}
    type="checkbox"
    value="foo"
  />
  <div
    checked={false}
    className="c2 c3"
  />
</label>
`;
